Minor end-to-end testing utilities

oskar-codes opened the pull request:
This pull request adds some minor utilities that aim to help running end-to-end tests locally. These are only helpful tools for developers and are not meant to be included in the CI. There are two Gradle commands that have been added:
- `startFirebaseEmulators` Simply starts the Firebase emulators as defined by `firebase.json` and imports the data stored in `firebase/emulator-data`.
- `connectedCheckWithFirebaseEmulators` Same as above, but also runs `connectedCheck` when the emulators are launched and closes the emulators after the tests are executed.

Note that I have only tested these Gradle commands on Windows as I only have a Windows PC. Please let me know if they do not work on Linux or MacOS.

This pull request also cleans up a little the `EndToEndTest` class, with better error handling and constant extraction.

Aurelien9Code commented on app/build.gradle.kts:
```diff
@@ -352,24 +352,44 @@ configurations.forEach { configuration ->
+        ).start()
+    } else if (os.contains("mac")) {
+        ProcessBuilder(
+            "osascript",
```
Thanks for using osascript to add support for macOS ; )

Aurelien9Code commented on app/build.gradle.kts:
```diff
@@ -352,24 +352,44 @@ configurations.forEach { configuration ->
+                gnome-terminal -- bash -c '$command; exec bash' ||
+                konsole -e '$command'
+                """.trimIndent()
+        ).start()
```
As you can see at the bottom of the [Java documentation for ProcessBuilder.start()](https://docs.oracle.com/javase/8/docs/api/index.html?java/lang/ProcessBuilder.html), the method can throw an `IOException`: "Starting an operating system process is highly system-dependent. Among the many things that can go wrong are:

    The operating system program file was not found.
    Access to the program file was denied.
    The working directory does not exist. 

In such cases an exception will be thrown. The exact nature of the exception is system-dependent, but it will always be a subclass of [IOException](https://docs.oracle.com/javase/8/docs/api/java/io/IOException.html).". Consider handling it with a `catch (e: IOException)`, and other errors with `catch (e: Exception)`.

oskar-codes replied:
Indeed, that's intentional :)
If an error occurs when running the command, it should be shown to the developper running it.

Aurelien9Code commented on app/build.gradle.kts:
```diff
@@ -352,24 +352,44 @@ configurations.forEach { configuration ->
 
+fun runCommand(command: String) {
+    val os = System.getProperty("os.name").toLowerCase()
+    if (os.contains("win")) {
```
Could be replaced with `when`conditional logic, as you already have three cases, and as it is better for fallbacks.

Aurelien9Code commented on app/src/androidTest/java/com/android/unio/end2end/EndToEndTest.kt:
```diff
@@ -62,19 +62,17 @@ open class EndToEndTest : FirebaseEmulatorFunctions {
     try {
-      Firebase.firestore.useEmulator(Firestore.HOST, 8080)
-      Firebase.auth.useEmulator(Auth.HOST, 9099)
+      Firebase.firestore.useEmulator(HOST, Firestore.PORT)
```
Well done ! Seems really more readable now : )

Aurelien9Code commented on app/src/androidTest/java/com/android/unio/end2end/EndToEndTest.kt:
```diff
@@ -62,19 +62,17 @@ open class EndToEndTest : FirebaseEmulatorFunctions {
 
               override fun onResponse(call: okhttp3.Call, response: okhttp3.Response) {
-                if (response.body == null) {
+                if (response.body == null || !response.body!!.string().contains("Ok")) {
```
Haha I have no idea how to change that but isn't there is a Json response ? That means we could check that the response is OK without consuming the response body. If not that seems really fine

oskar-codes replied:
The body isn't a JSON response, it is is simply the string `Ok` when the emulator is successfully running. Thus, it should be okay to consume the body as it will not be used anywhere else.

Aurelien9Code reviewed (approved):
Great job ! Well done for creating the runCommand function ! It's making the OS-specific logic reusable across multiple tasks ; )
Good job to check if the response body is OK.

Maybe handle more some possible exceptions as explained but overall it's perfect !  🚀

Zafouche commented on app/build.gradle.kts:
```diff
@@ -352,24 +352,44 @@ configurations.forEach { configuration ->
+                konsole -e '$command'
+                """.trimIndent()
+        ).start()
+    } else if (os.contains("mac")) {
+        ProcessBuilder(
+            "osascript",
+            "-e",
+            """
+                tell application "Terminal"
+                    do script "$command"
+                    activate
+                end tell
+                """.trimIndent()
+        ).start()
+    } else {
+        throw GradleException("Unsupported operating system: $os")
+    }
+}
```
It does not work unfortunately, it runs the script in 200ms, and although returning a successful result, it obviously does not run the tests.

Aurelien9Code reviewed (changes requested):
OK, in the end we tested on Linux with @armouldr  ! 
Now we know that it works. 
HOWEVER : everytime we restart the mac or the PC using linux, we need to run `chmod +x gradlew` in the root directory of the project. There wouldn't be so much of sense to this PR if instead of running a certain command we need to run a certain other - but only if you have a specfic OS. 
So I would advise just to do a chmod in root directory for linux and macos, before runing connectedCheck.

oskar-codes commented:
I believe the issue regarding execute permissions on `gradlew` is caused whenever you pull the file from Github, as it does not have the required permissions. We should consider pushing a `gradlew` file after granting execute permissions on it. 

Other than that, I have cleaned up the code and fixed an issue occurring on Windows, let me know when I can merge @Aurelien9Code.

Aurelien9Code commented:
> I believe the issue regarding execute permissions on `gradlew` is caused whenever you pull the file from Github, as it does not have the required permissions. We should consider pushing a `gradlew` file after granting execute permissions on it.
> 
> Other than that, I have cleaned up the code and fixed an issue occurring on Windows, let me know when I can merge @Aurelien9Code.

Okay I see ! I will push the gradlew updated with chmod access in my PR ; )
